librtt/librtt.c

```diff
@@ -0,0 +1,432 @@
+};
+
+
+struct rtt_desc {
+	char tag[16];
+	unsigned int txChannels;
+	unsigned int rxChannels;
+	struct rtt_pipe channel[LIBRTT_TXCHANNELS + LIBRTT_RXCHANNELS];
+};
+
+
+static const char librtt_tagReversed[] = LIBRTT_TAG_BACKWARD;
+static const size_t librtt_tagLength = sizeof(librtt_tagReversed) - 1;
+static const char *const librtt_txName[LIBRTT_TXCHANNELS] = LIBRTT_TXCHANNELS_NAMES;
+static const char *const librtt_rxName[LIBRTT_RXCHANNELS] = LIBRTT_RXCHANNELS_NAMES;
+
+
+static struct {
+	volatile struct rtt_desc *rtt;
+	librtt_cacheOp_t invalFn;
+	librtt_cacheOp_t cleanFn;
+	unsigned int lastRd[LIBRTT_TXCHANNELS];
+} librtt_common = { 0 };
+
+
+static void performCacheOp(librtt_cacheOp_t op, volatile unsigned char *buf, unsigned int sz, unsigned int rd, unsigned int wr)
+{
+	if ((rd == wr) || (op == NULL)) {
+		return;
+	}
+
+	if (wr < rd) {
+		op((void *)(buf + rd), sz - rd);
+		op((void *)buf, wr + 1);
+	}
+	else {
+		op((void *)(buf + rd), wr - rd + 1);
+	}
+}
+
+
+int librtt_checkTx(unsigned int ch)
+{
+	if ((librtt_common.rtt == NULL) || (ch >= librtt_common.rtt->txChannels)) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+
+int librtt_checkRx(unsigned int ch)
+{
+	if ((librtt_common.rtt == NULL) || (ch >= librtt_common.rtt->rxChannels)) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+
+ssize_t librtt_read(unsigned int ch, void *buf, size_t count)
+{
+	volatile struct rtt_desc *rtt = librtt_common.rtt;
+
+	if (librtt_checkRx(ch) < 0) {
+		return -ENODEV;
+	}
+
+	dataMemoryBarrier();
+	ch += rtt->txChannels;
+	volatile unsigned char *srcBuf = rtt->channel[ch].ptr;
+	unsigned char *dstBuf = (unsigned char *)buf;
+	unsigned int sz = rtt->channel[ch].sz - 1;
+	unsigned int rd = rtt->channel[ch].rd & sz;
+	unsigned int wr = rtt->channel[ch].wr & sz;
+	size_t todo = count;
+
+	performCacheOp(librtt_common.invalFn, (void *)srcBuf, sz + 1, rd, wr);
+	while ((todo != 0) && (rd != wr)) {
+		*dstBuf++ = srcBuf[rd];
+		rd = (rd + 1) & sz;
+		todo--;
+	}
+
+	rtt->channel[ch].rd = rd;
+
+	dataMemoryBarrier();
+
+	return count - todo;
+}
+
+
+ssize_t librtt_write(unsigned int ch, const void *buf, size_t count, int allowOverwrite)
+{
+	volatile struct rtt_desc *rtt = librtt_common.rtt;
+
+	if (librtt_checkTx(ch) < 0) {
+		return -ENODEV;
+	}
+
+	dataMemoryBarrier();
+
+	const unsigned char *srcBuf = (const unsigned char *)buf;
+	volatile unsigned char *dstBuf = rtt->channel[ch].ptr;
+	unsigned int sz = rtt->channel[ch].sz - 1;
+	unsigned int rd = (rtt->channel[ch].rd + sz) & sz;
+	unsigned int wr = rtt->channel[ch].wr & sz;
+	size_t todo = count;
+
+	while ((todo != 0) && (rd != wr)) {
+		dstBuf[wr] = *srcBuf++;
+		wr = (wr + 1) & sz;
+		todo--;
+	}
+
+	if ((rd == wr) && (allowOverwrite != 0)) {
+		while (todo != 0) {
+			dstBuf[wr] = *srcBuf++;
+			wr = (wr + 1) & sz;
+			todo--;
+		}
+
+		rtt->channel[ch].rd = wr;
+	}
+
+	performCacheOp(librtt_common.cleanFn, (void *)dstBuf, sz + 1, rd, wr);
+	rtt->channel[ch].wr = wr;
+
+	dataMemoryBarrier();
```

Comment:
This should actually be before the write to `wr` - if reorder happens and `wr` is set first, host/jlink may read data before librtt writes it.